source/common/nighthawk_sink_client_impl.cc

```diff
@@ -0,0 +1,56 @@
+#include "common/nighthawk_sink_client_impl.h"
+
+#include "external/envoy/source/common/common/assert.h"
+
+namespace Nighthawk {
+
+absl::StatusOr<nighthawk::client::StoreExecutionResponse>
+NighthawkSinkClientImpl::StoreExecutionResponseStream(
+    nighthawk::client::NighthawkSink::StubInterface* nighthawk_sink_stub,
+    const nighthawk::client::StoreExecutionRequest& store_execution_request) const {
+  ::grpc::ClientContext context;
+  ::nighthawk::client::StoreExecutionResponse store_execution_response;
+  std::shared_ptr<::grpc::ClientWriterInterface<::nighthawk::client::StoreExecutionRequest>> stream(
+      nighthawk_sink_stub->StoreExecutionResponseStream(&context, &store_execution_response));
+  if (!stream->Write(store_execution_request)) {
+    return absl::UnavailableError("Failed to write request to the Nighthawk Sink gRPC channel.");
+  } else if (!stream->WritesDone()) {
+    return absl::InternalError("WritesDone() failed on the Nighthawk Sink gRPC channel.");
+  }
+  ::grpc::Status status = stream->Finish();
+  if (!status.ok()) {
+    return absl::Status(static_cast<absl::StatusCode>(status.error_code()), status.error_message());
+  }
+  return store_execution_response;
+}
+
+absl::StatusOr<nighthawk::client::SinkResponse> NighthawkSinkClientImpl::SinkRequestStream(
+    nighthawk::client::NighthawkSink::StubInterface& nighthawk_sink_stub,
+    const nighthawk::client::SinkRequest& sink_request) const {
+  nighthawk::client::SinkResponse response;
+
+  ::grpc::ClientContext context;
+  std::shared_ptr<::grpc::ClientReaderWriterInterface<nighthawk::client::SinkRequest,
+                                                      nighthawk::client::SinkResponse>>
+      stream(nighthawk_sink_stub.SinkRequestStream(&context));
+
+  if (!stream->Write(sink_request)) {
+    return absl::UnavailableError("Failed to write request to the Nighthawk Sink gRPC channel.");
+  } else if (!stream->WritesDone()) {
+    return absl::InternalError("WritesDone() failed on the Nighthawk Sink gRPC channel.");
+  }
+
+  bool got_response = false;
+  while (stream->Read(&response)) {
+    RELEASE_ASSERT(!got_response,
```

Comment:
Why did we choose an ASSERT as compared to returning not-ok status? Is crashing the best choice here?

Reply:
This is something that really should not ever happen unless we explicitly decided otherwise. Having said that, personally I would lean slightly towards returning not ok and then cover this in a test as a behavioural aspect; but for now I settled for this because I think consistency is outweighing this (as this mirrors  another part of the code base).

Reply:
Happy to leave this up to you. For the sake of future discussions - this may not be a strong case for consistency. Consistency across the codebase is very important on items like APIs, used libraries, namespaces, etc. Individual functions should be able to make their own choice of whether they crash the code or return an error. I have learned to lean towards error returning, since crashes are harder to document and are often surprising to someone who might end up reusing the library or the function. Additionally crashes are not the best experience for our users.

With that said, I agree with you in this case that since we are making a choice to only return a single message, returning multiple is technically a programming error which does deserve a crash.